src/shared/sol-conffile.c

```diff
@@ -58,93 +61,198 @@ sol_conffile_entry_sort_cb(const void *data1, const void *data2)
+                entry->type = dup_string_from_json_token(value);
+            }
 
-        if (!entry->type) {
-            SOL_DBG("could not find mandatory 'Type' key on group [%s],"
-                " skipping entry", *groups_ptr);
-            free((void *)entry->id);
-            free(entry);
-            continue;
+            if (sol_json_token_str_eq(&key, node_options, strlen(node_options))) {
+                sol_conffile_set_entry_options(entry, value);
+            }
         }
 
-        entry->options = (const char **)g_key_file_get_string_list(keyfile,
-            *groups_ptr, "Options", NULL, NULL);
-
-        sol_ptr_vector_insert_sorted(&pv, entry,
-            sol_conffile_entry_sort_cb);
+        sol_ptr_vector_insert_sorted(&pv, entry, sol_conffile_entry_sort_cb);
     }
-    g_strfreev(groups);
-
     return pv;
 }
 
 static void
-sol_conffile_get_keyfile_include_paths(GKeyFile *keyfile,
+sol_conffile_get_keyfile_include_paths(
+    struct sol_json_scanner json_scanner,
     char **include,
     char **include_fallbacks)
 {
     const char *include_group = "SolettaInclude";
-    GError *error = NULL;
+    const char *include_str = "Include";
+    const char *include_fallback = "IncludeFallbacks";
 
-    if (!g_key_file_has_group(keyfile, include_group))
-        return;
+    bool found_include = false;
 
-    *include = g_key_file_get_string(keyfile,
-        include_group, "Include", &error);
-    if (*include == NULL && error
-        && error->code != G_KEY_FILE_ERROR_KEY_NOT_FOUND) {
-        SOL_DBG("could not read value from %s group: %s",
-            include_group, error->message);
-        g_error_free(error);
-    }
+    struct sol_json_token token, key, value;
+    struct sol_json_scanner include_scanner;
+    enum sol_json_loop_reason reason;
 
-    *include_fallbacks = g_key_file_get_string(keyfile, include_group,
-        "IncludeFallbacks", &error);
-    if (*include_fallbacks == NULL && error
-        && error->code != G_KEY_FILE_ERROR_KEY_NOT_FOUND) {
-        SOL_DBG("could not read value from %s group: %s",
-            include_group, error->message);
-        g_error_free(error);
+    SOL_JSON_SCANNER_OBJECT_LOOP (&json_scanner, &token, &key, &value, reason) {
+        if (sol_json_token_str_eq(&key, include_group, strlen(include_group))) {
+            found_include = true;
+            break;
+        }
     }
 
-    if (!g_key_file_remove_group(keyfile, include_group, &error)) {
-        SOL_DBG("could not remove %s group: %s", include_group,
-            error ? error->message : "unknown error");
-        g_error_free(error);
+    if (!found_include)
+        return;
+
+    sol_json_scanner_init_from_token(&include_scanner, &value);
+    SOL_JSON_SCANNER_OBJECT_LOOP (&include_scanner, &token, &key, &value, reason) {
+        if (sol_json_token_str_eq(&key, include_str, strlen(include_str))) {
+            *include = malloc(sol_json_token_get_size(&value) + 1);
```

Comment:
`strndup()`? Also, `malloc()` might fail. To make your life easier, a `sol_arena` might be used.